appsec/src/helper/sampler.hpp

```diff
@@ -6,42 +6,285 @@
+                // was it a hit?
+                if (cur_data.last_accessed == cur_data.last_reported) {
+                    // then this one should not be a hit
+                    return false;
+                }
+
+                // the other thread did not register a hit
+                // we retry if our idea of time is ahead
+                if (cur_data.last_accessed < now) {
+                    goto retry;
+                }
+            }
+
             return true;
+        } else {
+            // we just update the last accessed time
+            typename entry::edata desired_data{now, cur_data.last_reported};
+
+            // this would only be slightly innacurate:
+            // entry.data.store(desired_data, std::memory_order_relaxed);
+
+        retry2:
+            if (cur_data.last_accessed >= now) {
+                // we're behind the times
+                return false;
+            }
+
+            if (!entry.data.compare_exchange_strong(
+                    cur_data, desired_data, std::memory_order_relaxed)) {
+                goto retry2;
+            }
+
+            return false;
+        }
+    }
+
+public:
+    class table_accessor : private rcu_reader_state<table> {
+    public:
+        table_accessor(timed_set &set)
+            : rcu_reader_state<table>{set.rcu}, set{set}
+        {}
+
+        bool hit(std::uint64_t number) noexcept
+        {
+            auto guard{this->lock()};
+            auto &table = *guard;
+            return set.hit(table, number);
+        }
+
+        // for testing
+        auto approx_size() noexcept
+        {
+            auto guard{this->lock()};
+            return guard->size.load(std::memory_order_relaxed);
+        }
+
+    private:
+        timed_set &set;
+    };
+    friend class table_accessor;
+
+    ~timed_set()
+    {
+        // we must wait for the update thread
+        bool expected = false;
+        while (rebuild_in_progress.compare_exchange_strong(
+            expected, true, std::memory_order_relaxed)) {
+            std::this_thread::yield();
+        }
+    }
+
+    table_accessor new_accessor() { return table_accessor{*this}; }
+
+    std::unique_ptr<table_accessor> new_accessor_up()
+    {
+        return std::make_unique<table_accessor>(*this);
+    }
+
+    std::size_t approx_size(table &table) const { return table.size.load(); }
+
+private:
+    void rebuild_table(table &old_table, std::uint32_t report_threshold)
+    {
+        auto new_table = std::make_unique<table>();
+
+        struct copiable_entry {
+            std::uint64_t key;
+            typename entry::edata data;
+        };
+        struct comp_entry {
+            bool operator()(
+                const copiable_entry &lhs, const copiable_entry &rhs)
+            {
+                return lhs.data.last_accessed < rhs.data.last_accessed;
+            }
+        };
+
+        // most recent at the top
+        std::priority_queue<copiable_entry, std::vector<copiable_entry>,
```

Comment:
I think it would be faster to insert into a vector and then sort, since you don't need the data to be continuously sorted, you're incurring the cost of reordering the array on insertion.

Reply:
data in a heap is not fully sorted, and we don't need to sort all elements; we only need the top ones up to a certain number. That said, in these circumstances it's not necessarily faster than sorting the whole list at the end.

I'll look into the preallocation; it seems to require moving a preallocated vector into a different constructor overlaod.